### ISSUE TITLE
feat(model): native DFlash-2 block draft engine with parallel proposal generation

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12727]
# feat(model): native DFlash-2 block draft engine with parallel proposal generation

```routing
lane: internal/model
paths: internal/model/dflash2.go, internal/model/dflash2_test.go
expected_steps: 4
```

## Lane

`internal/model` — inference runtime, proposal generators, and speculative decoding algorithms.

## Problem frame

- Centrality: Core
- Portfolio tier: 2 (serving)
- Priority: P1
- P1: advanced - generates multi-token speculative blocks in a single non-autoregressive pass, eliminating serial draft step latency
- P2: preserved - preserves 100% of target model KV cache capacity by eliminating external sidecar KV pool collapse
- P3: advanced - provides dynamic block confidence thresholding to prune volatile draft suffix positions
- P4: preserved - integrates into SpeculativeEngine as a standard pluggable ProposalGenerator

## Parent context

Part of the Native Speculative Acceleration Series (`docs/tickets/native-speculative-acceleration/INDEX.md`), introducing high-throughput block-level draft generation.

## Current state

In standard speculative decoding with sidecar draft models (`internal/model/speculative.go:DraftModelProposalGenerator`), proposing $K$ candidate tokens requires running $K$ sequential autoregressive forward steps through the draft model (`session.Step(tok)` in a loop). Even for a small 1.5B or 3B draft model, $K$ sequential steps accumulate significant kernel launch overhead, memory bandwidth demand, and latency.

Furthermore, empirical studies of external sidecar drafters (such as standard multi-layer transformer drafters) reveal severe KV pool collapse: allocating a separate full KV cache for an external drafter starves the target model of addressable KV pages (e.g. collapsing target KV capacity from 104k to 15k tokens). DFlash-2 addresses this by utilizing a lightweight block-level draft architecture that predicts blocks of tokens in parallel conditioned on context representations without maintaining a separate massive KV cache.

## Why this is next

Autoregressive serial drafting bottlenecks speculative decoding when target model forward passes are fast (such as on unified memory APUs or high-bandwidth GPUs). By generating an entire block of 4 to 8 speculative tokens in a single non-autoregressive forward step, DFlash-2 cuts proposal latency by 3x to 4x while completely avoiding draft KV memory bloat.

## Working spine

`DFlash2ProposalGenerator` receives committed context tokens -> retrieves target model embedding/hidden representation -> executes compact DFlash-2 block projection head -> predicts block of $B$ candidate tokens in parallel with associated per-token confidence scores -> prunes low-confidence tail tokens below threshold -> packages candidate block into `DraftProposal` -> dispatches to `TargetVerificationEngine`.

## Core through-line

1. Implement `DFlash2ProposalGenerator` adhering to `ProposalGenerator` in `internal/model/dflash2.go`.
2. Define the DFlash-2 block prediction architecture: a compact 1–2 layer feed-forward and cross-attention block head that accepts target context features and generates $B$ output token logits in parallel ($B \in [4, 8]$).
3. Implement confidence-based block truncation: compute entropy or max-probability confidence per block slot, truncating the proposed block at the first token whose confidence falls below a configurable threshold (e.g. 0.35) to prevent proposing high-entropy tail noise.
4. Eliminate external KV cache allocations: operate DFlash-2 generation using static scratchpad buffers without reserving independent KV cache pages from the system pool.
5. Register DFlash-2 with `SpeculativeEngine`, supporting both standalone block speculation and hybrid cascading with prompt n-gram heuristics.
6. Provide unit and regression tests verifying valid proposal generation, confidence gating behavior, and zero divergence during target model verification.

## Gold-plating boundary

Strictly excluded from this work unit:
- Training or fine-tuning DFlash-2 block weights (uses pre-trained checkpoint weights or MTP conversion).
- Multi-block recurrent diffusion iterations beyond single-pass block drafting.
- Architecture-specific GPU assembly micro-optimizations outside standard compute tensor operations.

## Verifiable Witness

```bash
go test -v ./internal/model -run 'TestDFlash2ProposalGenerator' -count=1
```

## Witness

```bash
go test -v ./internal/model -run 'TestDFlash2ProposalGenerator' -count=1
```

### File:Line Seam

- `internal/model/speculative.go:20 (ProposalGenerator)`
- `internal/model/dflash2.go:1 (DFlash2ProposalGenerator)`
- `internal/model/speculative_verify.go:1 (TargetVerificationEngine)`

### Blast Radius & Affected Lanes

- **Primary Lane:** `internal/model`.
- **Unaffected Lanes:** `internal/compute`, `internal/agent`, `platform/*` remain unaffected.

### Quarantined Fallback Mechanism

If DFlash-2 block weights are not loaded or confidence scores drop below minimum bounds, the engine falls back transparently to prompt n-gram proposals or single-token decoding with zero interruption to active requests `[SW-VERIFIED]`.

## Scoped Acceptance Criteria

- [ ] 1. [SW-VERIFIED] Implement `DFlash2ProposalGenerator` in `internal/model/dflash2.go` generating parallel candidate blocks of length 4–8 in a single forward pass.
- [ ] 2. [SW-VERIFIED] Implement confidence thresholding gating low-confidence tail positions without generating invalid token IDs.
- [ ] 3. [SW-VERIFIED] Unit tests assert zero separate KV cache page allocations for the drafter, preserving complete target KV capacity.
- [ ] 4. [SW-VERIFIED] Target model verification tests verify exact output token identity when using DFlash-2 proposals under temperature-zero decoding.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that DFlash-2 generates 5 draft tokens in < 1.5ms proposal time, achieving >= 1.8x net end-to-end decode speedup.

## Done condition

- [ ] 1. [SW-VERIFIED] Implement `DFlash2ProposalGenerator` in `internal/model/dflash2.go` generating parallel candidate blocks of length 4–8 in a single forward pass.
- [ ] 2. [SW-VERIFIED] Implement confidence thresholding gating low-confidence tail positions without generating invalid token IDs.
- [ ] 3. [SW-VERIFIED] Unit tests assert zero separate KV cache page allocations for the drafter, preserving complete target KV capacity.
- [ ] 4. [SW-VERIFIED] Target model verification tests verify exact output token identity when using DFlash-2 proposals under temperature-zero decoding.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that DFlash-2 generates 5 draft tokens in < 1.5ms proposal time, achieving >= 1.8x net end-to-end decode speedup.

## Acceptance gate

`go test -v ./internal/model -run 'TestDFlash2ProposalGenerator' -count=1` exits 0 with 0 boundary violations.

## Closure binding

Resolving commit carries `(fak model)` trailer.

## Likely files

- `internal/model/dflash2.go`
- `internal/model/dflash2_test.go`
- `internal/model/speculative.go`
- `internal/model/speculative_verify.go`

## Expected steps

4
